doc/developer/design/20231102_read_holds.md

```diff
@@ -0,0 +1,154 @@
+# Read Holds
+
+- Durable Subscribes epic: [17541](https://github.com/MaterializeInc/materialize/issues/17541)
+- Configurable compaction windows: [16701](https://github.com/MaterializeInc/materialize/issues/16701)
+
+## The Problem
+
+Users are not able to use `SUBSCRIBE` without data loss.
+When materialize restarts or a client connection fails for some reason,
+users are not able to restart the `SUBSCRIBE` `AS OF` their most recently observed timestamp because the read frontier has usually progressed past that time.
+This happens because the compaction window is 1 second for all user objects.
+
+## Success Criteria
+
+`SUBSCRIBE` must be usable without data loss across server restarts or client reconnections.
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+## Solution Proposal
+
+Add a new `HOLD` object that acquires controller read holds on listed objects at a specific timestamp.
+The `HOLD` must be manually advanced to newer timestamps by users which will release the previous and acquire a new read hold to the specified timestamp.
+If the timestamp falls behind a configurable window, the `HOLD` is marked in an error state and the read holds are released.
+
+The `HOLD` can be created by:
+
+```
+CREATE HOLD <name>
+ON <object1>[, <object2> ...]
+[AT <time>]
+[WITH (
+    ERROR_IF_OLDER_THAN = '24h'::INTERVAL
+)]
+```
+
+A `HOLD` can be created on indexes, materialized views, tables, and sources.
+It cannot exist on sinks or unmaterialized views.
+If a `HOLD` exists for a directly queryable object (i.e., not indexes) at time `t`, `SELECT` and `SUBSCRIBE` on that object `AS OF t` will successfully execute.
+If a `HOLD` exists for an index, the same is true if the index satisfies the query.
+A `HOLD` on any non-index object will not also create the `HOLD` on those indexes.
+Because this may be surprising to users, if an `AS OF` query whose plan uses an index errors because the `AS OF` is less than the index's read frontier and a `HOLD` exists on the index's object,
+a helpful error message will be produced telling the user that their `HOLD` must be on the index, not the underlying object.
```

Comment:
An alternative here is that an `AS OF` query should restrict the indexes it considers using to those that are valid for the requested time, and we should offer to plan around those indexes if they are not valid. Idk what we do at the moment (probably not that) but wanted to float that there is another behavior we could consider.

Reply:
+1, I think this would be ideal as an end state, but in the meantime I think an error is sufficient.

Reply:
What if you do if there are two available indexes, one with a valid hold and one without? ISTM that we have to write the code to only consider the indexes with valid holds no matter what, and at that point using _no_ index if all indexes are invalid for the time seems to fall out naturally.

Reply:
I think it means that `AS OF` queries that don't find an index readable at the requested time, but find an underlying object (table, source) that is readable at that time, they should be planned to use that object (table, source) instead.

Which seems like a logical thing to do, but also like a potential footgun, if users didn't intend to query persist data directly. If the (table, source) is huge querying it directly might OOM the replica.

Reply:
> Which seems like a logical thing to do, but also like a potential footgun, if users didn't intend to query persist data directly. If the (table, source) is huge querying it directly might OOM the replica.

Yup, but the problem's not necessarily something we need to solve with the read holds work! Like, that problem exists today. You can be in the wrong cluster, or have built the wrong index, and get surprised. Possibly what we want is a session variable or a bit of query syntax that lets you say "produce an error if this can't use an index." E.g., imagine `SELECT ONLY FROM INDEX ... FROM ...`.

Reply:
Wait, I still see a sentence about the error message—is there a commit you haven't pushed up yet?

Reply:
I just forgot to remove the error message sentence. Removed.